src/Symfony/Component/Translation/Bridge/Crowdin/CrowdinProvider.php

```diff
@@ -0,0 +1,397 @@
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Translation\Bridge\Crowdin;
+
+use Psr\Log\LoggerInterface;
+use Symfony\Component\Translation\Dumper\XliffFileDumper;
+use Symfony\Component\Translation\Exception\ProviderException;
+use Symfony\Component\Translation\Loader\LoaderInterface;
+use Symfony\Component\Translation\Provider\ProviderInterface;
+use Symfony\Component\Translation\TranslatorBag;
+use Symfony\Component\Translation\TranslatorBagInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+
+/**
+ * @author Andrii Bodnar <andrii.bodnar@crowdin.com>
+ *
+ * In Crowdin:
+ *  * Filenames refers to Symfony's translation domains;
+ *  * Identifiers refers to Symfony's translation keys;
+ *  * Translations refers to Symfony's translated messages
+ *
+ * @experimental in 5.3
+ */
+final class CrowdinProvider implements ProviderInterface
+{
+    /** @var HttpClientInterface */
+    private $client;
+
+    /** @var LoaderInterface */
+    private $loader;
+
+    /** @var LoggerInterface */
+    private $logger;
+
+    /** @var XliffFileDumper */
+    private $xliffFileDumper;
+
+    /** @var string */
+    private $defaultLocale;
+
+    /** @var string */
+    private $endpoint;
+
+    /** @var string */
+    private $projectId;
+
+    /** @var array */
+    private $files = [];
+
+    public function __construct(HttpClientInterface $client, LoaderInterface $loader, LoggerInterface $logger, XliffFileDumper $xliffFileDumper, string $defaultLocale, string $endpoint, int $projectId)
+    {
+        $this->client = $client;
+        $this->loader = $loader;
+        $this->logger = $logger;
+        $this->xliffFileDumper = $xliffFileDumper;
+        $this->defaultLocale = $defaultLocale;
+        $this->endpoint = $endpoint;
+        $this->projectId = $projectId;
+    }
+
+    public function __toString(): string
+    {
+        return sprintf('crowdin://%s', $this->endpoint);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function write(TranslatorBagInterface $translatorBag): void
+    {
+        foreach ($translatorBag->getCatalogues() as $catalogue) {
+            foreach ($catalogue->getDomains() as $domain) {
+                $content = $this->xliffFileDumper->formatCatalogue($catalogue, $domain, ['default_locale' => $this->defaultLocale]);
+
+                $fileId = $this->getFileId($domain);
+
+                if ($catalogue->getLocale() === $this->defaultLocale) {
+                    if (!$fileId) {
+                        $this->addFile($domain, $content);
+                    } else {
+                        $this->updateFile($fileId, $domain, $content);
+                    }
+                } else {
+                    $this->uploadTranslations($fileId, $domain, $content, $catalogue->getLocale());
```

Comment:
The way this provider is implemented with private methods encapsulating the HTTP requests entirely means that the loop over domains and catalogues cannot benefit from the built-in concurrency features of symfony/http-client (as you block on the response methods before starting other request)